lib/active_model/serializer/adapter/json.rb

```diff
@@ -4,44 +4,122 @@ class ActiveModel::Serializer::Adapter::Json < ActiveModel::Serializer::Adapter
+            serialized_hash = add_resource_relationships(serialized_hash, serializer)
+          end
+
+          { root => serialized_hash }
+        end
+
+        def fragment_cache(cached_hash, non_cached_hash)
+          ActiveModel::Serializer::Adapter::Json::FragmentCache.new().fragment_cache(cached_hash, non_cached_hash)
+        end
+
+        private
 
-            core = cache_check(serializer) do
-              serializer.attributes(options)
+        # iterate through the associations on the serializer,
+        # adding them to the parent as needed (as singular or plural)
+        #
+        # nested_associations is a list of symbols that governs what
+        # associations on the passed in seralizer to include
+        def add_resource_relationships(parent, serializer, nested_associations = {})
+          # have the include array normalized
+          nested_associations = ActiveModel::Serializer::Utils.include_args_to_hash(nested_associations)
+
+          included_associations = if nested_associations.present?
+                                    serializer.associations.select { |association|
+                                      # nested_associations is a hash of:
+                                      #   key => nested association to include
+                                      nested_associations.key?(association.name)
+                                    }
+                                  else
+                                    serializer.associations
+                                  end
+
+          included_associations.each do |association|
+            serializer = association.serializer
+            opts = association.options
+            key = association.key
+
+            # sanity check if the association has nesting data
+            has_nesting = nested_associations[key].present?
+            if has_nesting
+              include_options_from_parent = { include: nested_associations[key] }
+              opts = opts.merge(include_options_from_parent)
             end
 
-            serializer.associations.each do |association|
-              serializer = association.serializer
-              opts = association.options
-
-              if serializer.respond_to?(:each)
-                array_serializer = serializer
-                hash[association.key] = array_serializer.map do |item|
-                  cache_check(item) do
-                    item.attributes(opts)
-                  end
-                end
-              else
-                hash[association.key] =
-                  if serializer && serializer.object
-                    cache_check(serializer) do
-                      serializer.attributes(options)
-                    end
-                  elsif opts[:virtual_value]
-                    opts[:virtual_value]
-                  end
-              end
+            if serializer.respond_to?(:each)
+              parent[key] = add_has_many_relationship(serializer, opts)
```

Comment:
Maybe we could rename this to `has_many_relationships_for(serializer, opts)` to keep it consistent with the JSONAPI adapter naming?